src/mod_mam_muc_elasticsearch_arch.erl

```diff
@@ -0,0 +1,286 @@
+-export([archive_size/4]).
+
+-include("mongoose.hrl").
+-include("mongoose_rsm.hrl").
+-include("mod_mam.hrl").
+-include("jlib.hrl").
+
+-define(INDEX_NAME, <<"muc_messages">>).
+-define(TYPE_NAME, <<"muc">>).
+
+%%-------------------------------------------------------------------
+%% gen_mod callbacks
+%%-------------------------------------------------------------------
+
+-spec start(jid:server(), list()) -> ok.
+start(Host, _Opts) ->
+    ejabberd_hooks:add(hooks(Host)),
+    ok.
+
+-spec stop(jid:server()) -> ok.
+stop(Host) ->
+    ejabberd_hooks:delete(hooks(Host)),
+    ok.
+
+%%-------------------------------------------------------------------
+%% ejabberd_gen_mam_archive callbacks
+%%-------------------------------------------------------------------
+
+archive_message(_Result, Host, MessageId, _UserId, RoomJid, _SourceJid, SourceJid, _Dir, Packet) ->
+    Room = mod_mam_utils:bare_jid(RoomJid),
+    SourceBinJid = mod_mam_utils:full_jid(SourceJid),
+    DocId = make_document_id(Room, MessageId),
+    Doc = make_document(MessageId, Room, SourceBinJid, Packet),
+    case mongoose_elasticsearch:insert_document(?INDEX_NAME, ?TYPE_NAME, DocId, Doc) of
+        {ok, _} ->
+            ok;
+        {error, _} = Err ->
+            ?ERROR_MSG("Failed to archive message in ElasticSearch: ~p", [Err]),
+            ejabberd_hooks:run(mam_drop_message, Host, [Host]),
+            Err
+    end.
+
+lookup_messages(_Result, _Host, Params) ->
+    SearchQuery0 = build_search_query(Params),
+    Sorting = [#{mam_id => #{order => determine_sorting(Params)}}],
+    ResultLimit = maps:get(page_size, Params),
+    SearchQuery1 = SearchQuery0#{sort => Sorting,
+                                 size => ResultLimit},
+    SearchQuery2 = maybe_add_from_constraint(SearchQuery1, Params),
+    case mongoose_elasticsearch:search(?INDEX_NAME, ?TYPE_NAME, SearchQuery2) of
+        {ok, Result} ->
+            {ok, search_result_to_mam_lookup_result(Result, Params)};
+        {error, _} = Err ->
+            ?ERROR_MSG("Failed to lookup messages in ElasticSearch: ~p", [Err]),
+            Err
+    end.
+
+-spec archive_size(Size :: integer(),
+                   Host :: jid:server(),
+                   _ArchiveId,
+                   RoomJid :: jid:jid()) -> non_neg_integer().
+archive_size(_Size, _Host, _ArchiveId, RoomJid) ->
+    SearchQuery = build_search_query(#{owner_jid => RoomJid}),
+    archive_size(SearchQuery).
+
+-spec remove_archive(Acc,
+                     Host :: jid:server(),
+                     ArchiveId :: mod_mam:archive_id(),
+                     RoomJid :: jid:jid()) -> Acc when Acc :: map().
+remove_archive(Acc, _Host, _ArchiveId, RoomJid) ->
+    SearchQuery = build_search_query(#{owner_jid => RoomJid}),
+    case mongoose_elasticsearch:delete_by_query(?INDEX_NAME, ?TYPE_NAME, SearchQuery) of
+        ok ->
+            ok;
+        {error, _} = Err ->
+            ?ERROR_MSG("Failed to delete messages in ElasticSearch: ~p", [Err]),
```

Comment:
Where is that documented? Is there a logging standard?

Reply:
And how could I know that the `event` has to be `remove_archive_failed` for that particular action?

Reply:
Not really documented.
Just grep for "event="

```erlang
git grep "event="

src/ejabberd_auth_odbc.erl:            ?ERROR_MSG("event=check_password_failed "                                                  
src/ejabberd_auth_odbc.erl:            ?ERROR_MSG("event=check_password_failed "                                                  
src/ejabberd_auth_odbc.erl:            ?ERROR_MSG("event=check_password_failed "                                                  
src/ejabberd_auth_odbc.erl:            ?ERROR_MSG("event=check_password_failed "                                                  
src/ejabberd_auth_odbc.erl:            ?ERROR_MSG("event=try_register_failed "                                                    
src/ejabberd_auth_odbc.erl:            ?ERROR_MSG("event=get_vh_registered_users_failed "                                         
src/ejabberd_auth_odbc.erl:        ?ERROR_MSG("event=get_vh_registered_users_failed "                                             
src/ejabberd_auth_odbc.erl:            ?ERROR_MSG("event=get_vh_registered_users_failed "                                         
src/ejabberd_auth_odbc.erl:        ?ERROR_MSG("event=get_vh_registered_users_failed "                                             
src/ejabberd_auth_odbc.erl:            ?ERROR_MSG("event=get_vh_registered_users_numbers_failed "                                 
src/ejabberd_auth_odbc.erl:        ?ERROR_MSG("event=get_vh_registered_users_numbers_failed "                                     
src/ejabberd_auth_odbc.erl:            ?ERROR_MSG("event=get_vh_registered_users_numbers_failed "                                 
src/ejabberd_auth_odbc.erl:            ?ERROR_MSG("event=get_password_failed "                                                    
src/ejabberd_auth_odbc.erl:            ?ERROR_MSG("event=get_password_s_failed "                                                  
src/ejabberd_auth_odbc.erl:            ?ERROR_MSG("event=does_user_exist_failed "                                                 
src/ejabberd_auth_odbc.erl:            ?ERROR_MSG("event=does_user_exist_failed "                                                 
src/ejabberd_auth_odbc.erl:        ?ERROR_MSG("event=remove_user_failed "                                                         
src/ejabberd_c2s.erl:            ?ERROR_MSG("event=conflict_check_failed " 
```

there are still two formats:

```erlang
key1=value1, key2=value2

and

key1=value1 key2=value2
```

The old discussion https://github.com/esl/MongooseIM/issues/1248

After that we've decided to use `event=something_happening` format instead of `msg=something_happening` or `issue=something_happening`.

Reply:
Added some docs about logging in a PR https://github.com/esl/MongooseIM/pull/1906